solutions/AC Matching.mdx

```diff
@@ -0,0 +1,57 @@
+---
+id: ac-matching
+source: AtCoder DP Contest
+title: Matching
+author: Neo Wang
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
+$S$ to the first $|S|$ males, this problem boils down to the following
+formulation:
+
+$$
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
+$$
+
+In English, this is the equivalent of
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
+> equivalent to the sum of all the matchings without female $x$ such that the
+> addition of female $x$ is compatible with the $|S|$-th male.
+
+Our base cases are built using matches of the first male; if the first male is
+compatible with the $i$-th female, then $\{F_i\}$ will have one match.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\cdot 2^N)$
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+bool compat[21][21];
+int dp[1 << 21];
+
+int N;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	cin >> N;
+	for(int i = 0; i < N; i++) for(int j = 0; j < N; j++) cin >> compat[i][j];
+
+	for(int i = 0; i < N; i++) if(compat[0][i]) dp[1 << i] = 1;
+	for(int s = 1; s < (1<<N); s++) {
```

Comment:
Can't you set `dp[0]=1` and start the loop at `s=0`?

Reply:
yeah that works too.